src/Symfony/Component/ObjectMapper/ObjectMapper.php

```
@@ -0,0 +1,226 @@
+        private readonly ?PropertyAccessorInterface $propertyAccessor = null,
+        private readonly ?ContainerInterface $serviceLocator = null
+    ) {
+    }
+
+    public function map(object $object, object|string $to = null): object
+    {
+        static $objectMap = null;
+        $objectMapInitialized = false;
+
+        if (null === $objectMap) {
+            $objectMap = new \SplObjectStorage();
+            $objectMapInitialized = true;
+        }
+
+        $refl = new \ReflectionClass($object);
+
+        $metadata = $this->metadataFactory->create($object);
+        $map = $this->getMapTo($metadata, null, $object);
+        $to ??= $map?->to;
+
+        if (!(\is_string($to) && class_exists($to)) && !\is_object($to)) {
+            throw new RuntimeException(sprintf('Attribute of type "%s" expected on "%s.', Map::class, $refl->getName()));
+        }
+
+        $toRefl = new \ReflectionClass($to);
+        $mapped = \is_object($to) ? $to : $toRefl->newInstanceWithoutConstructor();
+        if ($map && $map->transform) {
+            $mapped = $this->applyTransforms($map, $mapped, $mapped);
+        }
+
+        if (!\is_object($mapped)) {
+            throw new RuntimeException('Can not map to a non-object.');
+        }
+
+        $objectMap[$object] = $mapped;
+
+        $arguments = [];
+        $constructor = $toRefl->getConstructor();
+        foreach ($constructor?->getParameters() ?? [] as $parameter) {
+            $parameterName = $parameter->getName();
+            $property = $toRefl->getProperty($parameterName);
+
+            // The mapped class was probably instantiated in a transform we can't write a readonly property
+            if ($property->isReadOnly() && ($property->isInitialized($mapped) && $property->getValue($mapped))) {
+                continue;
+            }
+
+            $arguments[$parameterName] = $parameter->isDefaultValueAvailable() ? $parameter->getDefaultValue() : null;
```

Not sure about the null value here, IMO it should throw a specific exception when there is a required parameter and we could not find a value from the source object

---

I think that the thrown exception will be explicit enough (thrown by `$constructor?->invokeArgs($mapped, $arguments);`). Indeed `arguments` may find their value when we loop over properties later on.